File: app/policies/admin_policy.rb
class AdminPolicy < ApplicationPolicy
  def index?
    return user.super? || user.owner? if admin?

    false
  end

  def show?
    return user.super? || user.owner? || record == user if admin?

    false
  end

  def roles?
    return user.super? || user.owner? if admin?

    false
  end

  def create?
    return user.super? || user.owner? if admin?

    false
  end

  def update?
    return user.super? || user.owner? if admin?

    false
  end

  def suspend?
    return user.super? || user.owner? if admin?

    false
  end

  class Scope < Scope
    def resolve
      scope
    end
  end
end

File: app/javascript/src/shared/policies/admin_policy.js
import PolicyBase from './policy_base'

export default class AdminPolicy extends PolicyBase {
  constructor(user, record = null) {
    super(user, record)
  }

  index() {
    return this.forAdminOwnerAndSuperOnly()
  }

  show() {
    return this.forAdminOwnerAndSuperOrSelfOnly()
  }

  roles() {
    return this.forAdminOwnerAndSuperOnly()
  }

  create() {
    return this.forAdminOwnerAndSuperOnly()
  }

  update() {
    return this.forAdminOwnerAndSuperOnly()
  }

  suspend() {
    return this.forAdminOwnerAndSuperOnly()
  }

  // helpers

  forAdminOwnerAndSuperOrSelfOnly() {
    if (this.isAdmin()) {
      return this.user.isOwner() || this.user.isSuper() || this.record.id === this.user.id
    } else {
      return false
    }
  }
}
